### DISCUSSION TITLE
libass rendering is problematic when {\p\bord}

### DISCUSSION THREAD

**Discussion by op200** [arch1t3cht/Aegisub#146] [General]
![图片](https://github.com/arch1t3cht/Aegisub/assets/112297583/dad3af83-a626-409e-8a7b-6b746baa2e7f)
Aegisub's built-in libass
This bug does not exist in VSFilter  
  
![图片](https://github.com/arch1t3cht/Aegisub/assets/112297583/48ea55fa-eba4-43f0-ba1c-7d887ea50df2)
This image was rendered by MPC-HC's libass

**Comment by arch1t3cht** [comment]
It's hard to see what exactly you're referring to here without seeing the full subtitle file, but this is probably because libass applies stroking to shapes while vsfilter applies it to the bitmaps. The rendering differentials arising from this are a known issue in libass (see e.g. libass/libass#302, libass/libass#123, libass/libass#582), but it seems quite possible that this won't ever be fixed. Either way, this is something that should be reported at libass (but doesn't need to be, since it's known already), not here, since Aegisub has no control over this.

**Reply by op200** [reply]
What I want to express is that this bug appears in Aegisub's built-in libass, but I found that MPC-HC's built-in libass filter and 'XySubFilter with libass v3.0.0.209 20230505 with libass 0.17.1' do not have it
So I think this could be an issue with Aegisub, or it could be a problem with the built-in version of libass

```
{\bord1}{\p1}m 0 0 l 200 0 l 200 70 l 0 70 {\p0}
```

**Reply by 0tkl** [reply]
(Reply to op200) Can't reproduce. 

![image](https://github.com/arch1t3cht/Aegisub/assets/118708188/2d09e629-8e5a-4448-ae93-23bff19e92fc)

If you have a FULL subtitle file, I could test it on a dummy video.

**Reply by op200** [reply]
> (Reply to 0tkl) it need {\1aFF}
```
[Script Info]
; Script generated by Aegisub 9706-cibuilds-20caaabc0
; http://www.aegisub.org/
Title: Default Aegisub file
ScriptType: v4.00+
WrapStyle: 0
ScaledBorderAndShadow: yes
YCbCr Matrix: None
PlayResX: 1920
PlayResY: 1080

[Aegisub Project Garbage]
Video File: ?dummy:23.976:38000:1920:1080:47:163:254:c
Video AR Value: 1.777778
Video Zoom Percent: 0.875000

[V4+ Styles]
Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding
Style: Default,Arial,48,&HFFFFFFFF,&H000000FF,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2,0,2,10,10,10,1

[Events]
Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text
Dialogue: 0,0:00:00.00,0:00:05.00,Default,,0,0,0,,{\bord1}{\p1}m 0 0 l 800 0 l 800 800 l 0 800 {\p0}
```

**Reply by 0tkl** [reply]
Now it can be confirmed. I don't know why, but on my Aegisub, the inconsistent width appears on some random sizes, although not the same as the size you provided.

![image](https://github.com/arch1t3cht/Aegisub/assets/118708188/27b2b636-e4ee-4cb9-89a6-d99fbd4d5437)

![image](https://github.com/arch1t3cht/Aegisub/assets/118708188/e09b7c73-0eed-4b27-9011-8a370e3993fa)

Also you mentioned MPC-HC, so i tried it on my local mpv. The rendering problem doesn't appear on mpv either.

```
PS C:\Users\username\AppData\Local\mpv> .\mpv --version
mpv v0.38.0-580-g00f43e09 Copyright © 2000-2024 mpv/MPlayer/mplayer2 projects
 built on Jul  7 2024 00:04:00
libplacebo version: v7.349.0 (v7.349.0-dirty)
FFmpeg version: N-116159-gecd3a9783
FFmpeg library versions:
   libavcodec      61.9.100
   libavdevice     61.2.100
   libavfilter     10.2.102
   libavformat     61.4.100
   libavutil       59.28.100
   libswresample   5.2.100
   libswscale      8.2.100
```

**Reply by op200** [reply]
> (Reply to 0tkl)
So I guess this is due to Aeigsub, or due to the version of libass

**Reply by arch1t3cht** [reply]
I see what you mean now, very interesting. For what it's worth, this also happens on my version of mpv (on Windows):

> mpv v0.37.0-522-ga56d8ff1 Copyright © 2000-2024 mpv/MPlayer/mplayer2 projects
 built on Mar  9 2024 10:22:56
libplacebo version: v7.347.0 (v6.338.0-102-g0d3542d-dirty)
FFmpeg version: N-114081-g64634e809
FFmpeg library versions:
   libavutil       59.1.100
   libavcodec      61.1.100
   libavformat     61.0.100
   libswscale      8.0.100
   libavfilter     10.0.100
   libswresample   5.0.100

![image](https://github.com/arch1t3cht/Aegisub/assets/99741385/8f2ed88b-a95f-44ad-9ae3-57368ea46019)

**Reply by op200** [reply]
Now it is confirmed that the bug is caused by libass scaling
Thank you for your answers
